### ISSUE TITLE
Fix (or don't) the behavior of calling "Focus" on Android to open the picker

### ISSUE THREAD

**Issue by PureWeen** [dotnet/maui#8946]
Currently the behavior of calling "Focus" to open the picker dialogs is broken on Android because of this [PR](https://github.com/dotnet/maui/pull/8913).  That behavior has always been a bit of a hack and should most likely be replaced by [this](https://github.com/dotnet/maui/issues/8945). That being said it does complicate the migration story to remove this behavior. 

We could tie a focus call at the controls level into the APIs we add for https://github.com/dotnet/maui/issues/8945. 

### Reproduction

```XAML
<DatePicker Visible="false" x:name="picker" />
```

```C#
picker.Focus()
```

**Comment by RemkoVermaeren** [comment]
Did you find any workaround to open the picker on focus?

**Comment by planyazilim** [comment]
I hope there will be solution about this bug soon. Almost every App uses DatePicker and I am sure many of troubling this issue.

**Comment by ralphmagnette** [comment]
Is there already a solution or workaround for this bug, need to deploy my application on both Apple Store and Google Play?

**Comment by Ghostbird** [comment]
[Related discussion](https://github.com/dotnet/maui/discussions/7410#8946). It's annoying that Github doesn't automatically link these like it does for issues and PRs.

**Comment by nonoodev** [comment]
> ion o

The **very dirty** fix we found was to replace
`<Picker IsVisible="False x:Name="ItemPicker" ...... />`
by
`<Picker Opacity="0" HeightRequest="1" WidthRequest="1" x:Name="ItemPicker" ...... />`

and update our OpenPicker method as following :
```
private void OpenPicker(object sender, EventArgs e)
        { 
            ItemPicker.Unfocus();
            ItemPicker.Focus();
            // TODO Update when bug will be fixed
        }
```
(we have a TapGestureRecognizer that call the OpenPicker method)

I don't think it's a good way to fix it for production but we had to make it work quickly to continue the development of the app and it's working.

**Comment by Ghostbird** [comment]
@nonoodev Can you explain _how_ this accomplishes opening the picker? As far as I know the issue is that the code that opens the picker has been removed from the `Focus()` method in #8913.

From what I can see, you had an `ItemPicker` that was not rendered (`IsVisible="False"`), and now you render it transparent. Any tapped picker will open by default. An non-rendered picker is not tappable, but a transparent picker can be tapped. To me it seems as if you code has nothing to do with this issue, and executes some unnecessary code as an aside.

**Comment by nacompllo** [comment]
I have found a way to open the date picker, Android's native View class (Android.Views) have a method called PerformClick which allows opening the date picker without pressing the DatePicker and TimePicker, I leave you an example of how I did it in case it can help someone in the same situation.

I have made a small example in which I have a DatePicker, when the user clicks on it the date picker appears, when we select a date from the picker and click "OK" the DateSelected event is fired and this is where the magic is done to access the TimePicker Handler and thus be able to access the native "PerformClick" method to open the TimePicker time selector without having to click on it:

**XAML:**
```
        <StackLayout
            HorizontalOptions="CenterAndExpand"
            VerticalOptions="CenterAndExpand">
            <DatePicker 
                x:Name="datePicker"
                DateSelected="datePicker_DateSelected"
                BackgroundColor="White"
                TextColor="Black"
                FontSize="20"
                WidthRequest="200"
                HeightRequest="200">
            </DatePicker>

            <TimePicker 
                x:Name="timePicker"
                IsVisible="False"/>
        </StackLayout>
```

**Code-behind:**
```
    private void datePicker_DateSelected(object sender, DateChangedEventArgs e)
    {
#if ANDROID
        var handler = timePicker.Handler as ITimePickerHandler;
        handler.PlatformView.PerformClick();
#endif
    }
```

**Image:**
![fixFocusIssue](https://user-images.githubusercontent.com/20904914/221552532-fb8d45e0-c5aa-43e7-af7a-622c0198120c.gif)

**Comment by echolumaque** [comment]
Here we go again folks :)

**Comment by YakushenkoES** [comment]
Is there any solution for Windows?

**Comment by JimmyPun610** [comment]
Still an issue in .NET8.
https://github.com/dotnet/maui/issues/8946#issuecomment-1446178827 works but there is another problem.
Since the focus and unfocus event do not fire at all, there is no way to do any action on dialog onclose

**Comment by kelputoo123** [comment]
They really need to fix this, i can't go further in my app without this thing.
I will try to use a different datepicker control from syncfusion or something like that, must work.

**Comment by SERGIOLNA** [comment]
> Still an issue in .NET8. [#8946 (comment)](https://github.com/dotnet/maui/issues/8946#issuecomment-1446178827) works but there is another problem. Since the focus and unfocus event do not fire at all, there is no way to do any action on dialog onclose

Yes, you can detect the closing of the DatePicker dialog, not directly with the unfocus event, the MAUI abstraction layer as it currently stands does not allow it, but you can create a handler and access the native closing events of the DatePicker in In the case of Android I think they are Cancel (handles the cancellation/closure event) and Dismiss (handles when the dialog is closed in any way), if you are not familiar with the handlers Maui still has support for custom renderer in case it is easier for you .

**Comment by nacompllo** [comment]
> Still an issue in .NET8. [#8946 (comment)](https://github.com/dotnet/maui/issues/8946#issuecomment-1446178827) works but there is another problem. Since the focus and unfocus event do not fire at all, there is no way to do any action on dialog onclose

Yes, you can detect the closing of the DatePicker dialog, not directly with the unfocus event, the MAUI abstraction layer as it currently stands does not allow it, but you can create a handler and access the native closing events of the DatePicker in In the case of Android I think they are Cancel (handles the cancellation/closure event) and Dismiss (handles when the dialog is closed in any way), if you are not familiar with the handlers Maui still has support for custom renderer in case it is easier for you .

**Comment by Akinpellumie** [comment]
> I have found a way to open the date picker, Android's native View class (Android.Views) have a method called PerformClick which allows opening the date picker without pressing the DatePicker and TimePicker, I leave you an example of how I did it in case it can help someone in the same situation.
> 
> I have made a small example in which I have a DatePicker, when the user clicks on it the date picker appears, when we select a date from the picker and click "OK" the DateSelected event is fired and this is where the magic is done to access the TimePicker Handler and thus be able to access the native "PerformClick" method to open the TimePicker time selector without having to click on it:
> 
> **XAML:**
> 
> ```
>         <StackLayout
>             HorizontalOptions="CenterAndExpand"
>             VerticalOptions="CenterAndExpand">
>             <DatePicker 
>                 x:Name="datePicker"
>                 DateSelected="datePicker_DateSelected"
>                 BackgroundColor="White"
>                 TextColor="Black"
>                 FontSize="20"
>                 WidthRequest="200"
>                 HeightRequest="200">
>             </DatePicker>
> 
>             <TimePicker 
>                 x:Name="timePicker"
>                 IsVisible="False"/>
>         </StackLayout>
> ```
> 
> **Code-behind:**
> 
> ```
>     private void datePicker_DateSelected(object sender, DateChangedEventArgs e)
>     {
> #if ANDROID
>         var handler = timePicker.Handler as ITimePickerHandler;
>         handler.PlatformView.PerformClick();
> #endif
>     }
> ```
> 
> **Image:** ![fixFocusIssue](https://user-images.githubusercontent.com/20904914/221552532-fb8d45e0-c5aa-43e7-af7a-622c0198120c.gif) [ ![fixFocusIssue](https://user-images.githubusercontent.com/20904914/221552532-fb8d45e0-c5aa-43e7-af7a-622c0198120c.gif) ](https://user-images.githubusercontent.com/20904914/221552532-fb8d45e0-c5aa-43e7-af7a-622c0198120c.gif) [ ](https://user-images.githubusercontent.com/20904914/221552532-fb8d45e0-c5aa-43e7-af7a-622c0198120c.gif)

This works for me even to trigger datepicker.
For both iOS and Android, see below implementation

```
#if ANDROID
        var handler = datePicker.Handler as IDatePickerHandler;
        handler.PlatformView.PerformClick();
#endif

#if IOS
        var iosHandler = datePicker.Handler as IDatePickerHandler;
        iosHandler.PlatformView.ResignFirstResponder();
#endif
```

**Comment by Ghostbird** [comment]
> I think for a11y reasons, we cannot/should not open the picker. If you are navigating the screen using TalkBack, then suddenly it goes from a form to a date picker and then you are lost.
> 
> This must be a separate API where the dev can decide when/if to show the picker.

Agreed. My use case currently is only a date+time picker where I show the time picker when a date is picked.

**Comment by Ducvu1996** [comment]
Is there already a solution for this bug,? It is still not working on Android.

**Comment by ramonB1996** [comment]
Right now, the workaround to still show the picker on Android seems to be: 

In your View, give an `x:Name="picker"` to your Picker element.

In code-behind:

```
#if ANDROID
			var handler = picker.Handler as IPickerHandler;
			handler?.PlatformView.PerformClick();
#endif
```

This also works the same for DatePicker (IDatePickerHandler) and TimePicker (ITimePickerHandler).

**Comment by arunjose696** [comment]
Workaround by @Akinpellumie seems to work for me as well. Is there someway to trigger PerformClick() from MAUI as crossplatform?

In the [MAUI docs](https://learn.microsoft.com/en-us/dotnet/maui/user-interface/controls/picker?view=net-maui-9.0) it is mentioned "When the [Picker](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.picker) gains focus, its data is displayed and the user can select an item:" 

So would that mean this is a contradiction to docs as the focus does not bring up the data?

**Comment by kubaflo** [comment]
A maintainer explicitly rejected opening the picker through Focus for accessibility reasons and required a separate API instead.

Closing as not planned because this is an explicit maintainer or product decision, not an inference from age or inactivity.
